aggie-coding-club/Rev-Registration
# Store courses in sessions

## Description

Saves courses when generate schedules is clicked, and sets courses on loading the page or choosing a term.

## Rationale

I had to change the `courseCards` redux pretty substantially to make everything work, so if there are any changes you guys want to suggest then I'm open to it. The main thing that's a bit strange with it is that `replaceCourseCards` doesn't actually create an action, it just dispatches several different actions depending on the arguments. The reason for this is so that all the intermediate states can be shown instead of having to calculate everything and then updating state, but I can change it to do everything simultaneously.

I also changed the formatting of the `sessions/get_last_term` API to return empty json instead of an empty string if a last term doesn't exist since it's simpler, didn't require any frontend changes, and is consistent with how the `get_saved_courses` API works.

Last thing is that I fetch updated sections and update the saved course cards since the seats/available sections, etc might have changed since the schedule was generated, I tried to keep as much information as possible from the saved schedules though. It might be worthwhile to simplify how schedules are represented on the backend (instead of serializing the entire `courseCards` redux slice, only store certain parts to reduce storage requirements) because not all of the information is used, but for now I'm keeping it simple. 

## How to test

Do whatever you want with course cards, click "generate schedules", and refresh/change terms, etc., and the state of your course cards should be exactly the same.

## Related tasks

Related to #260, still need to save availabilities but courses are pretty complicated as it involves fetching/modifying data rather than direct serialization/deserialization so it's not happening in this PR
